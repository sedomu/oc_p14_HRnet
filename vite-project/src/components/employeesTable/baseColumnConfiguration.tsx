import type { Column } from "@tanstack/react-table";
import type { Employee } from "@/redux.ts";
import { Button } from "@/components/ui/button.tsx";
import { getSortingIcon } from "@/components/employeesTable/getSortingIcon.tsx";

export default function baseColumnConfiguration(
    accessorKey: string,
    displayName: string
) {
    return {
        accessorKey: accessorKey,
        header: ({ column }: { column: Column<Employee> }) => (
            <Button
                variant={"ghost"}
                className="bg-transparent hover:bg-transparent hover:text-inherit"
                onClick={() =>
                    column.toggleSorting(column.getIsSorted() === "asc")
                }
            >
                {displayName}
                {getSortingIcon(column)}
            </Button>
        ),
    };
}
