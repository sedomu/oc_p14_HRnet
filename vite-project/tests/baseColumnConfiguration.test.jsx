import { describe, it, expect, vi } from "vitest";
import baseColumnConfiguration from "@/components/employeesTable/baseColumnConfiguration.js";
import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";

describe("Base column configuration", () => {
    it("should be able to configure a column", () => {
        const accessorKey = "testAccessorKey";
        const displayName = "testDisplayName";

        const result = baseColumnConfiguration(accessorKey, displayName);
        expect(result.accessorKey).toBe(accessorKey);
        expect(typeof result.header).toBe("function");

        const fakeColumn = {
            toggleSorting: vi.fn(),
            getIsSorted: vi.fn().mockReturnValue("asc"),
        };
        const { container } = render(
            <>{result.header({ column: fakeColumn })}</>
        );
        expect(container.textContent).toBe(displayName);

        expect(container).toMatchSnapshot();
    });
    it("should be able to launch sorting method when clicked", async () => {
        const toggleSorting = vi.fn();
        const getIsSorted = vi.fn().mockReturnValue("asc");

        const fakeColumn = { toggleSorting, getIsSorted };

        const config = baseColumnConfiguration(
            "testAccessorKey",
            "Test Column"
        );

        render(<>{config.header({ column: fakeColumn })}</>);

        const button = screen.getByText("Test Column");

        await userEvent.click(button);

        expect(toggleSorting).toHaveBeenCalledWith(true);
    });
});
