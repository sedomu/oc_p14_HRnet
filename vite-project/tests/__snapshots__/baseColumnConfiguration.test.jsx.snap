// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Base column configuration > should be able to configure a column 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:hover:bg-accent/50 h-9 px-4 py-2 has-[>svg]:px-3 bg-transparent hover:bg-transparent hover:text-inherit"
    data-slot="button"
  >
    testDisplayName
    <svg
      aria-hidden="true"
      class="lucide lucide-arrow-down ml-2 h-4 w-4"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 5v14"
      />
      <path
        d="m19 12-7 7-7-7"
      />
    </svg>
  </button>
</div>
`;
